### ISSUE TITLE
Events loss in Marten.Events.Projections.Async.Fetcher under load

### ISSUE THREAD

**Issue by simbadltd** [JasperFx/marten#1879]
## Issue
There is events loss under load in 3.x, e.g. several transactions appending event (minimum 2 concurrent transactions). The same issue has been described here: [ Async daemon can lose events in a highly concurrent scenario #1629 ](https://github.com/JasperFx/marten/issues/1629).

## Root Cause
When an event is added to an aggregate in our code, `mt_append_event` function has been called. This function opens new transaction inside the database. Each transaction allocate it's own unique `seq_id` from `mt_events_sequence`. Finish transactions order is not defined and not guaranteed by PostgreSQL. This means that transaction which started later may commit sooner. And that's why we can observe records with greater `seq_id` sooner than records with smaller `seq_id`. 

`Marten.Events.Projections.Async.Fetcher` in the process of fetching next events page makes a request like:
```sql
select seq_id from public.mt_events where seq_id > [LAST] 
and seq_id <= [LIMIT] 
order by seq_id
```
This means that if  `Fetcher` will pick larger `seq_id` (because its' transaction committed sooner), then events with smaller `seq_id` (because transaction has committed later) will be skipped because of `[LAST]` parameter for the next fetch will be updated to `seq_id` number from second transation.

## Solution
The main idea is to use PostgreSQL `tx_id` (transaction id). It is monotonically increasing number handling by PostgreSQL itself. This means that we can have unique increasing number for each event in correct order: the later event the smaller number. We also have `txid_snapshot_xmin(txid_current_snapshot())` construction which returns minimum `tx_id` (transaction id) of any opened (not finished, not committed) transaction. So, we are going to use it in way that we can chuck out all committed later than minimum not finished transaction events: `tx_id < txid_snapshot_xmin(txid_current_snapshot())`. Once, not finished transaction committed, all this events become visible for `Fetcher` and will be handled:
```sql
select seq_id from public.mt_events where seq_id > [LAST]
and seq_id <= [LIMIT]
and tx_id < txid_snapshot_xmin(txid_current_snapshot())
order by seq_id
```

`events` table and `mt_append_event` function also require changes:

```sql
DROP TABLE IF EXISTS {databaseSchema}.mt_events;
CREATE TABLE {databaseSchema}.mt_events (
	seq_id		bigint CONSTRAINT pk_mt_events PRIMARY KEY,
	id		uuid NOT NULL,
	stream_id	uuid REFERENCES {databaseSchema}.mt_streams ON DELETE CASCADE,
	version		integer NOT NULL,
	data		jsonb NOT NULL,
	type 		varchar(100) NOT NULL,
	timestamp	timestamptz default (now()) NOT NULL,
	tx_id bigint DEFAULT 0 NOT NULL,
	CONSTRAINT pk_mt_events_stream_and_version UNIQUE(stream_id, version),
	CONSTRAINT pk_mt_events_id_unique UNIQUE(id)
);
```

```sql
CREATE OR REPLACE FUNCTION {databaseSchema}.mt_append_event(stream uuid, stream_type varchar, event_ids uuid[], event_types varchar[], bodies jsonb[]) RETURNS int[] AS $$
DECLARE
	event_version int;
	event_type varchar;
	event_id uuid;
	body jsonb;
	index int;
	seq int;
	return_value int[];
BEGIN
	select version into event_version from {databaseSchema}.mt_streams where id = stream;
	if event_version IS NULL then
		event_version = 0;
		insert into {databaseSchema}.mt_streams (id, type, version, timestamp) values (stream, stream_type, 0, now());
	end if;


	index := 1;
	return_value := ARRAY[event_version + array_length(event_ids, 1)];

	foreach event_id in ARRAY event_ids
	loop
	    seq := nextval('{databaseSchema}.mt_events_sequence');
		return_value := array_append(return_value, seq);

	    event_version := event_version + 1;
		event_type = event_types[index];
		body = bodies[index];

		insert into {databaseSchema}.mt_events
			(seq_id, id, stream_id, version, data, type, tx_id)
		values
			(seq, event_id, stream, event_version, body, event_type, txid_current());



		index := index + 1;
	end loop;

	update {databaseSchema}.mt_streams set version = event_version, timestamp = now() where id = stream;

	return return_value;
END
$$ LANGUAGE plpgsql;
```

This solution should work for any version of PostgreSQL from 9.6 to 12.0. In 13.0 there was a breaking-change. But it's not a problem to support this solution in 13.0 too.

I'd just want to share my thoughts on how it is possible to solve such kind of weird behavior. And I'm going to use this fix in production, but first I'd like to know your opinion.

**Comment by oskardudycz** [comment]
@simbadltd, thank you for the fix! :) It's available in v3.13.5. See more:
- https://github.com/JasperFx/marten/releases/tag/3.13.5
- https://www.nuget.org/packages/Marten/3.13.5

**Comment by oskardudycz** [comment]
Fixed by #1629

**Comment by culaja** [comment]
I had the same problem with 3.13.14 using the following API:
```csharp
      private IEnumerable<List<Marten.Events.IEvent>> GetAllEventBatchesUsing(
                  IDocumentSession session,
                  long position)
      {
          return session.Events.QueryAllRawEvents()
              .Where(e => e.Sequence > position)
              .OrderBy(e => e.Sequence)
              .InBatches(_batchSize);
      }
```
Unfortunately, I still have the same problem after switching to 3.13.15 since QueryAllRawEvents will skip uncommitted transactions. How can I solve this problem using QueryAllRawEvents?

**Comment by oskardudycz** [comment]
@culaja, could you elaborate on the exact issue? Postgres by design does not support dirty reads. Thus, until events are committed, it's technically impossible to read them outside of the transaction that's appending them.

**Comment by culaja** [comment]
Bad description :)

QueryAllRawEvents behaves the same as Fetcher before the fix. It will not take into account
```sql
and tx_id < txid_snapshot_xmin(txid_current_snapshot())
```

In my opinion all API methods for fetching global events (outside of a specific stream) should always take into account tx_id < txid_snapshot_xmin

**Comment by Leh2** [comment]
@simbadltd we are using the fix now and it seems to run well. Especially because we can now lower the `LeadingEdgeBuffer` 👍 

You mentioned a breaking change in postgres 13. We are using postgres 13 so are there something we should be aware of? 

> In 13.0 there was a breaking-change. But it's not a problem to support this solution in 13.0 too.

**Comment by simbadltd** [comment]
@Leh2 Great news! I think you don't have to worry about any problems in postgres 13. But there may be problems when upgrading to new versions since `txid_` functions are deprecated in 13:

> In releases of PostgreSQL before 13 there was no xid8 type, so variants of these functions were provided that used bigint to represent a 64-bit XID, with a correspondingly distinct snapshot data type txid_snapshot. These older functions have txid in their names. They are still supported for backward compatibility, but may be removed from a future release.

For more information, please refer [this](https://www.postgresql.org/docs/13/functions-info.html#FUNCTIONS-TXID-SNAPSHOT).

**Comment by Leh2** [comment]
Cool thanks for info ❤️

**Comment by Leh2** [comment]
Hi @simbadltd & @oskardudycz 

I'm seeing results where the same event is processed twice. This doesn't happen consistently, but very often is our setup.

My best guess would be that the `txid_snapshot_xmin(txid_current_snapshot())` is changing between the following calls:

```sql
select seq_id from internal.mt_events where seq_id > 336391126 and seq_id <= 336391276 and extract(epoch from age(transaction_timestamp(), internal.mt_events.timestamp)) >= 0 and tx_id < txid_snapshot_xmin(txid_current_snapshot()) order by seq_id;
select * from internal.mt_events where seq_id > 336391126 and seq_id <= 336391276 and extract(epoch from age(transaction_timestamp(), internal.mt_events.timestamp)) >= 0 and tx_id < txid_snapshot_xmin(txid_current_snapshot()) order by seq_id;
select min(seq_id) from internal.mt_events where seq_id > 336391276 and extract(epoch from age(transaction_timestamp(), internal.mt_events.timestamp)) >= 0 and tx_id < txid_snapshot_xmin(txid_current_snapshot());
select max(seq_id) from internal.mt_events where seq_id >= 336391276 and extract(epoch from age(transaction_timestamp(), internal.mt_events.timestamp)) >= 0 and tx_id < txid_snapshot_xmin(txid_current_snapshot());
```

This simple example shows the problem:

```sql
select txid_snapshot_xmin(txid_current_snapshot());
select txid_snapshot_xmin(txid_current_snapshot());
select txid_snapshot_xmin(txid_current_snapshot());
select txid_snapshot_xmin(txid_current_snapshot());
```

->

| txid_snapshot_xmin | 
| ------------- | 
| 109824224  |
| 109824229  |
| 109824230 |
| 109824230 |

I'm not sure what the best solution would be to use the same `txid`, but it could be plpgsql, `WITH` or temp table. Do you have any good suggestions?

**Comment by simbadltd** [comment]
Hi @Leh2! It seems that there is nothing to do with `tx_id`, duplicate processing might be the case for previous versions too. If I understand correctly, [following LOC](https://github.com/JasperFx/marten/blob/76e00b7141fdb241a1a29db5b85a7ca28e482a76/src/Marten/Events/Projections/Async/Fetcher.cs#L211) can return `events` that are out of sync with the `sequences`. So, to avoid duplicate processing you need to cutoff events with `Sequence` > `sequences.Last()`. This is just a guess. Would you mind checking this hypothesis?

**Comment by Leh2** [comment]
> Hi @Leh2! It seems that there is nothing to do with `tx_id`, duplicate processing might be the case for previous versions too. If I understand correctly, [following LOC](https://github.com/JasperFx/marten/blob/76e00b7141fdb241a1a29db5b85a7ca28e482a76/src/Marten/Events/Projections/Async/Fetcher.cs#L211) can return `events` that are out of sync with the `sequences`. So, to avoid duplicate processing you need to cutoff events with `Sequence` > `sequences.Last()`. This is just a guess. Would you mind checking this hypothesis?

Sounds reasonable, will give it a shot 🔫

**Comment by Leh2** [comment]
Hi again @simbadltd. I found a couple scenarios where we could get into trouple:

The following is true for both scenarios:

`PageSize`: 10
`LastEncountered`: 0
`Limit` (0+10): 10

**Scenario 1: Event(s) is processed twice when event(s) is becoming available between sequences and events fetch:**

_-> 5 available events_
Fetch sequences: 0..5
_-> New Events is available_
Fetch events: 0..6

`NextKnownSequence` and `LastKnownSequence` will both be 0 as we aren't hitting our limit.

For the next run the `LastEncountered` (based on sequences) will only be 5 and the last event is processed twice. 

**Scenario 2: Let's say we fixed the above scenario by using `sequences.Last()` as our truth**

_-> 9 available events_
Fetch sequences: 0..9
_-> New Events is available_
Fetch events: 0..10, but will be truncated to 0..9 by our new fix
`NextKnownSequence seq_id > :limit`: 0
`LastKnownSequence seq_id >= :limit`: 10

For the next run the `LastEncountered` (based on `LastKnownSequence`) will be 10 and we will miss an event. 

I'm not sure what the best solution would be as we also need to consider scenarios with big gabs of ignore events and therefore can't use `sequences` as the only truth as that would slow down the processing. 

By best take for a solution would be to query for sequence limit to start with before running the existing queries:

`select max(seq_id) from mt_events where seq_id > :last and extract(epoch from age(transaction_timestamp(), mt_events.timestamp)) >= :buffer and tx_id < txid_snapshot_xmin(txid_current_snapshot())`

This result should be our new baseline for existing events and ensures all queries are running on the same events without new events are coming in-between selects.

**Comment by simbadltd** [comment]
Hi @Leh2 ! I'm not sure that I understand the `Scenario 2` correctly. But as I see the query for sequences is executed first. So, it won't be the case. The initial problem here is that the query for sequences and query for events (actually payload) are executed in different transactions. That's why we can get more events than seq-ids. And that's why we need to cut off all events with `seq_id` greater than `sequences.Last()`.

Anyway, we can refine `LastKnownSequence ` as well.

**Comment by Leh2** [comment]
The scenario 2 is about the last query `select max(seq_id) ... where seq_id >= :limit ...` for `LastKnownSequence`.

We missed an event in the sequence query and therefore removed the last events from the event query, but because `LastKnownSequence` got a value this will be used for the [LastEncountered()](https://github.com/JasperFx/marten/blob/3.14/src/Marten/Events/Projections/Async/EventPage.cs#L64-L75) method. 

`Sequence` query is working on `seq_id <= :limit` and `LastKnownSequence` is working on ` seq_id >= :limit` so i'm not sure how this should be refined?

The `LastEncountered` logic was introduced [here](https://github.com/JasperFx/marten/compare/64e9d36be410253b7431e38dae6f93059d6712f3...88cd64bd5a8d07a725f6b8f33834661382b81990) and should be a fix for big gap problem raised [here](https://github.com/JasperFx/marten/issues/624).

**Comment by simbadltd** [comment]
> `Sequence` query is working on `seq_id <= :limit` and `LastKnownSequence` is working on ` seq_id >= :limit` so i'm not sure how this should be refined?
> 

@Leh2 I mean that we can use `sequences.Last()` if any sequence is present. Instead of calculating `LastEncountered` using magic numbers [here](https://github.com/JasperFx/marten/blob/76e00b7141fdb241a1a29db5b85a7ca28e482a76/src/Marten/Events/Projections/Async/EventPage.cs#L71). I see that it was so initially, but then 70f7bfda27a397a8d07dd67175a3189de183be46 fix added logic mentioned above. May be the initial problem could be fixed by introducing a guarantee of sequential event processing (like we do using `tx_id`). That's why I suggest to try rely on the `sequences.Last()` first.

**Comment by Leh2** [comment]
Guarantee of sequential event processing should certainly help as long we still account for gaps caused by rolled backed transactions.

It's sounds like you have a good idea how to approach this, how you like to give a try? ❤️

**Comment by simbadltd** [comment]
@Leh2 Sorry for the delay. Until the 70f7bfda27a397a8d07dd67175a3189de183be46 was added, the sequences were taken into account as I described, but then for some reason the logic was changed. And there are some unit-tests which contradict with my idea. I'll think about scenario #2 some more. But it seems that it can be a conceptual problem of the AsyncFetcher architecture. And `tx_id` filter doesn't change the behavior.
